gk/main.c

```diff
@@ -157,17 +154,33 @@ priority_from_delta_time(uint64_t present, uint64_t past)
 	return integer_log_base_2(delta_time);
 }
 
+static int
+get_grantor_id(struct gk_lpm *ltbl, struct ip_flow *flow)
+{
+	if (flow->proto == ETHER_TYPE_IPv4)
+		return lpm_lookup_ipv4(ltbl->lpm, flow->f.v4.dst);
+
+	if (likely(flow->proto == ETHER_TYPE_IPv6))
+		return lpm_lookup_ipv6(ltbl->lpm6, flow->f.v6.dst);
+
+	rte_panic("Unexpected condition at %s: unknown flow type %hu\n",
+		__func__, flow->proto);
+
+	return -1; /* Unreachable. */
+}
+
 static inline void
-initialize_flow_entry(struct flow_entry *fe, struct ip_flow *flow)
+initialize_flow_entry(struct flow_entry *fe,
+	struct ip_flow *flow, struct gk_fib *grantor_fib)
 {
 	rte_memcpy(&fe->flow, flow, sizeof(*flow));
 
 	fe->state = GK_REQUEST;
 	fe->u.request.last_packet_seen_at = rte_rdtsc();
 	fe->u.request.last_priority = START_PRIORITY;
 	fe->u.request.allowance = START_ALLOWANCE - 1;
-	/* TODO Grantor ID comes from LPM lookup. */
-	fe->u.request.grantor_id = 0;
+	fe->grantor_fib = grantor_fib;
+	grantor_fib->ref_cnt++;
```

Comment:
It's better to leave the TODO comment saying that the reference is not being added because the solution requires more attention that just adding the increment above. The following comment is a good start:

```
/*
 * TODO Flow entries should maintain the reference counter of
 * the grantor FIB entry to avoid the entry to go away before the flow.
 *
 * Notice that all GK blocks are calling this function, so a solution needs
 * to deal with concurrency.
 *
 * Moreover, the chose solution must be very efficient due to the impact on
 * the throughout of the GK blocks. For example, a simple atomic counter will
 * slow down all GK blocks; especially if there is only one grantor FIB entry.
 *
 * Ideas for solution: trade the need to maintain the reference counter of
 * the grantor FIB entry for some expensive operation that is only needed
 * while editing the FIB table.
 */
```